domain/infosync/resource_group_manager.go

```diff
@@ -0,0 +1,218 @@
+// limitations under the License.
+
+package infosync
+
+import (
+	"context"
+	"sync"
+
+	"github.com/pingcap/errors"
+	rmpb "github.com/pingcap/kvproto/pkg/resource_manager"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tidb/parser/model"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+// ConvertToProtoResourceGroup converts model.ResourceGroupInfo to rmpb.ResourceGroup
+func ConvertToProtoResourceGroup(modelGroup *model.ResourceGroupInfo) (*rmpb.ResourceGroup, error) {
+	group := &rmpb.ResourceGroup{
+		Name:     modelGroup.Name.String(),
+		Settings: &rmpb.GroupSettings{},
+	}
+	var isRUMode bool
+	if modelGroup.RRURate > 0 || modelGroup.WRURate > 0 {
+		isRUMode = true
+		group.Settings.Mode = rmpb.GroupMode_RUMode
+		group.Settings.RUSettings = &rmpb.GroupRequestUnitSettings{
+			RRU: &rmpb.TokenBucket{
+				Settings: &rmpb.TokenLimitSettings{
+					Fillrate: modelGroup.RRURate,
+				},
+			},
+			WRU: &rmpb.TokenBucket{
+				Settings: &rmpb.TokenLimitSettings{
+					Fillrate: modelGroup.WRURate,
+				},
+			},
+		}
+	}
+	if len(modelGroup.CPULimiter) > 0 || len(modelGroup.IOReadBandwidth) > 0 || len(modelGroup.IOWriteBandwidth) > 0 {
+		if isRUMode {
+			return nil, errors.New("Only one of RUMode and NativeMode can be set")
+		}
+		parseF := func(s string, scale resource.Scale) (uint64, error) {
+			if len(s) == 0 {
+				return 0, nil
+			}
+			q, err := resource.ParseQuantity(s)
+			if err != nil {
+				return 0, err
+			}
+			return uint64(q.ScaledValue(scale)), nil
+		}
+		cpuRate, err := parseF(modelGroup.CPULimiter, resource.Milli)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ioReadRate, err := parseF(modelGroup.IOReadBandwidth, resource.Scale(0))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ioWriteRate, err := parseF(modelGroup.IOWriteBandwidth, resource.Scale(0))
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		group.Settings.Mode = rmpb.GroupMode_NativeMode
+		group.Settings.ResourceSettings = &rmpb.GroupResourceSettings{
+			Cpu: &rmpb.TokenBucket{
+				Settings: &rmpb.TokenLimitSettings{
+					Fillrate: cpuRate,
+				},
+			},
+			IoRead: &rmpb.TokenBucket{
+				Settings: &rmpb.TokenLimitSettings{
+					Fillrate: ioReadRate,
+				},
+			},
+			IoWrite: &rmpb.TokenBucket{
+				Settings: &rmpb.TokenLimitSettings{
+					Fillrate: ioWriteRate,
+				},
+			},
+		}
+		return group, nil
+	}
+	if isRUMode {
+		return group, nil
+	}
+	return nil, errors.New("at least one setting in the group")
+}
+
+// ResourceGroupManager manages resource group settings
+type ResourceGroupManager interface {
+	// GetResourceGroup is used to get one specific rule bundle from ResourceGroup Manager.
+	GetResourceGroup(ctx context.Context, name string) (*rmpb.ResourceGroup, error)
+	// GetAllResourceGroups is used to get all rule bundles from ResourceGroup Manager.
+	GetAllResourceGroups(ctx context.Context) ([]*rmpb.ResourceGroup, error)
```

Comment:
How many resource groups there will be?
In the past we load all ddl history in one API and that cause OOM.

Reply:
A good point. Is there common approach at tidb side to follow, such as serializing to disk etc, to mitigate the problem?

Reply:
We expect up to 10k resource groups in one cluster in the extreme case, if the metadata of 1 resource group consumes 1k memory, seems this is not a big problem even in this extreme case

Reply:
OK